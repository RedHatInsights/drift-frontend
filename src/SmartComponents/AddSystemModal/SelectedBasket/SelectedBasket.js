import React, { Component } from 'react';
import PropTypes from 'prop-types';
import { Button, Popover, PopoverPosition } from '@patternfly/react-core';
import SelectedTable from './SelectedTable/SelectedTable';
import EmptyStateDisplay from '../../EmptyStateDisplay/EmptyStateDisplay';
import addSystemModalHelpers from '../redux/helpers';

export class SelectedBasket extends Component {
    constructor(props) {
        super(props);

        this.state = {
            systemsToDeselect: [],
            baselinesToDeselect: [],
            hspsToDeselect: []
        };

        this.clearDeselected = () => {
            this.setState({ systemsToDeselect: []});
            this.setState({ baselinesToDeselect: []});
            this.setState({ hspsToDeselect: []});
        };
    }

    async componentDidMount() {
        addSystemModalHelpers.setContent({
            systems: this.props.systems,
            baselines: this.props.baselines,
            historicalProfiles: this.props.historicalProfiles
        }, this.props.handleSystemSelection, this.props.handleBaselineSelection, this.props.handleHSPSelection);
    }

    toggleBasket = () => {
        const { toggleBasketVisible } = this.props;

        this.clearDeselected();
        toggleBasketVisible();
    }

    applyChanges = async () => {
        const { baselinesToDeselect, hspsToDeselect, systemsToDeselect } = this.state;
        const { handleBaselineSelection, handleHSPSelection, selectBaseline, selectEntity, selectHistoricProfiles,
            selectedBaselineContent, selectedHSPContent, toggleBasketVisible } = this.props;

        toggleBasketVisible();

        if (baselinesToDeselect.length) {
            await selectBaseline(baselinesToDeselect, false, 'COMPARISON');
            handleBaselineSelection(selectedBaselineContent.filter(baseline => baselinesToDeselect.includes(baseline.id)), false);
        }

        if (hspsToDeselect.length) {
            hspsToDeselect.forEach(async function(hsp) {
                await handleHSPSelection(selectedHSPContent.find(hspContent => hspContent.id === hsp));
            });

            await selectHistoricProfiles(selectedHSPContent.filter(item => !hspsToDeselect.includes(item.id))
            .map(({ id }) => id));
        }

        systemsToDeselect.forEach(async function(system) {
            await selectEntity(system, false);
        });

        this.clearDeselected();
    };

    findType = (type, id) => {
        const { systemsToDeselect, baselinesToDeselect, hspsToDeselect } = this.state;
        let newArray;

        if (type === 'system') {
            newArray = this.toggleSelected(systemsToDeselect, id);
            this.setState({ systemsToDeselect: newArray });
        } else if (type === 'baseline') {
            newArray = this.toggleSelected(baselinesToDeselect, id);
            this.setState({ baselinesToDeselect: newArray });
        } else {
            newArray = this.toggleSelected(hspsToDeselect, id);
            this.setState({ hspsToDeselect: newArray });
        }
    }

    toggleSelected = (array, id) => {
        let newArray = [];

        if (array.includes(id)) {
            newArray = this.removeId(id, array);
        } else {
            newArray = [ ...array ];
            newArray.push(id);
        }

        return newArray;
    }

    removeId(id, array) {
        let newArray = [];
        array.forEach(function(item) {
            if (id !== item) {
                newArray.push(item);
            }
        });

        return newArray;
    }

    findSelected() {
        const { selectedBaselineContent, selectedHSPContent, selectedSystemContent } = this.props;
        let selectedCount = 0;

        if (selectedSystemContent.length) {
            selectedCount += selectedSystemContent.length + selectedBaselineContent.length + selectedHSPContent.length;
        } else {
            selectedCount += selectedBaselineContent.length + selectedHSPContent.length;
        }

        return selectedCount;
    }

    displayBodyContent = (isEmpty) => {
        const { entities, selectedSystemContent, selectedBaselineContent, selectedHSPContent } = this.props;
        let bodyContent;

        if (isEmpty) {
            bodyContent = <EmptyStateDisplay
                title='Nothing selected'
                text={ [ 'Select systems and baselines to compare.' ] }
            />;
        } else {
            bodyContent = <SelectedTable
                selectedBaselineContent={ selectedBaselineContent }
                entities={ entities }
                selectedHSPContent={ selectedHSPContent }
                findType={ this.findType }
                handleDeselect={ this.handleDeselect }
                selectedSystemContent={ selectedSystemContent }
            />;
        }

        return bodyContent;
    }

    render() {
        const { isVisible } = this.props;

        return (
            <React.Fragment>
                <span>
                    <Popover
                        id='selected-basket'
                        style={{ minWidth: '500px' }}
                        isVisible={ isVisible }
                        shouldClose={ () => this.toggleBasket() }
                        headerContent={ <div>Selected items ({ this.findSelected() })</div> }
                        footerContent={ <Button
                            key="confirm"
                            variant="primary"
                            onClick={ () => this.applyChanges() }
                            ouiaId="confirm-selected-basket-button"
                        >
                            Apply changes
                        </Button> }
                        position={ PopoverPosition.bottom }
                        bodyContent={ <div style={{ maxHeight: '500px', overflowY: 'auto' }}>
                            { this.displayBodyContent(this.findSelected() === 0) }
                        </div> }
                    >
                        <a onClick={ () => this.toggleBasket() }>
                            Selected ({ this.findSelected() })
                        </a>
                    </Popover>
                </span>
            </React.Fragment>
        );
    }
}

SelectedBasket.propTypes = {
    baselines: PropTypes.array,
    entities: PropTypes.object,
    handleBaselineSelection: PropTypes.func,
    handleHSPSelection: PropTypes.func,
    handleSystemSelection: PropTypes.func,
    historicalProfiles: PropTypes.array,
    isVisible: PropTypes.bool,
    selectBaseline: PropTypes.func,
    selectedBaselineContent: PropTypes.array,
    selectedHSPContent: PropTypes.array,
    selectedSystemContent: PropTypes.array,
    selectEntity: PropTypes.func,
    selectHistoricProfiles: PropTypes.func,
    systems: PropTypes.array,
    toggleBasketVisible: PropTypes.func
};

export default SelectedBasket;
