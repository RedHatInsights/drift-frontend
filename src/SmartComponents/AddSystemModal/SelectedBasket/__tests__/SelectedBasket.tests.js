/*eslint-disable camelcase*/
import React from 'react';
import { shallow } from 'enzyme';
import toJson from 'enzyme-to-json';
import { BlueprintIcon, ClockIcon, ServerIcon } from '@patternfly/react-icons';

import SelectedBasket from '../SelectedBasket';

describe('SelectedBasket', () => {
    let props;

    beforeEach(() => {
        props = {
            selectedBaselineContent: [
                { id: 'abcd1234', name: 'baseline', icon: <BlueprintIcon /> }
            ],
            selectedHSPContent: [
                {
                    id: 'ijkl9101',
                    captured_date: '2021-03-03T06:40:32+00:00',
                    icon: <ClockIcon />,
                    system_id: 'efgh5678'
                }
            ],
            selectedSystemContent: [
                { id: 'efgh5678', name: 'system', icon: <ServerIcon /> }
            ],
            handleBaselineSelection: jest.fn(),
            handleHSPSelection: jest.fn(),
            selectBaseline: jest.fn(),
            selectHistoricProfiles: jest.fn(),
            selectEntity: jest.fn(),
            toggleBasketVisible: jest.fn()
        };
    });

    it('should render empty', () => {
        props.selectedSystemContent = [];
        props.selectedBaselineContent = [];
        props.selectedHSPContent = [];

        const wrapper = shallow(
            <SelectedBasket
                { ...props }
            />
        );

        expect(toJson(wrapper)).toMatchSnapshot();
    });

    it('should not be visible with selected count', () => {
        const wrapper = shallow(
            <SelectedBasket
                { ...props }
            />
        );

        expect(toJson(wrapper)).toMatchSnapshot();
    });

    it('should toggle to visible with system, baseline and hsp', () => {
        const wrapper = shallow(
            <SelectedBasket
                { ...props }
            />
        );

        wrapper.find('a').simulate('click');
        expect(toJson(wrapper)).toMatchSnapshot();
    });

    it('should remove a single content piece', () => {
        const wrapper = shallow(
            <SelectedBasket
                { ...props }
            />
        );

        wrapper.setState({ systemsToDeselect: [ 'efgh5678' ]});
        wrapper.setState({ isVisible: true });

        wrapper.instance().onToggle();
        expect(props.selectEntity).toHaveBeenCalledWith('efgh5678', false);
        expect(props.selectBaseline).not.toHaveBeenCalled();
        expect(props.selectHistoricProfiles).not.toHaveBeenCalled();
    });

    it('should remove one of two hsps', () => {
        props.selectedHSPContent.push({
            id: 'mnop1121',
            captured_date: '2021-03-02T06:40:32+00:00',
            icon: <ClockIcon />,
            system_id: 'efgh5678'
        });
        const wrapper = shallow(
            <SelectedBasket
                { ...props }
            />
        );

        wrapper.setState({ hspsToDeselect: [ 'ijkl9101' ]});
        wrapper.setState({ isVisible: true });

        wrapper.instance().onToggle();
        expect(props.selectHistoricProfiles).toHaveBeenCalledWith([ 'mnop1121' ]);
    });

    it('should remove all selected', async () => {
        const wrapper = shallow(
            <SelectedBasket
                { ...props }
            />
        );

        wrapper.setState({
            systemsToDeselect: [ 'efgh5678' ],
            baselinesToDeselect: [ 'abcd1234' ],
            hspsToDeselect: [ 'ijkl9101' ]
        });
        wrapper.setState({ isVisible: true });

        await wrapper.instance().onToggle();
        await expect(props.selectBaseline).toHaveBeenCalledWith([ 'abcd1234' ], false, 'COMPARISON');
        await expect(props.handleBaselineSelection).toHaveBeenCalledWith(
            [{ id: 'abcd1234', name: 'baseline', icon: <BlueprintIcon /> }],
            false
        );
        await expect(props.selectHistoricProfiles).toHaveBeenCalledWith([]);
        await expect(props.handleHSPSelection).toHaveBeenCalledWith({
            id: 'ijkl9101',
            captured_date: '2021-03-03T06:40:32+00:00',
            icon: <ClockIcon />,
            system_id: 'efgh5678'
        });
        await expect(props.selectEntity).toHaveBeenCalledWith('efgh5678', false);
    });

    it('should add system to systemsToDeselect', () => {
        const wrapper = shallow(
            <SelectedBasket
                { ...props }
            />
        );

        wrapper.instance().findType('system', 'efgh5678');

        expect(wrapper.state('systemsToDeselect')).toEqual([ 'efgh5678' ]);
    });

    it('should add baseline to baselinesToDeselect', () => {
        const wrapper = shallow(
            <SelectedBasket
                { ...props }
            />
        );

        wrapper.instance().findType('baseline', 'abcd1234');

        expect(wrapper.state('baselinesToDeselect')).toEqual([ 'abcd1234' ]);
    });

    it('should add hsp to hspsToDeselect', () => {
        const wrapper = shallow(
            <SelectedBasket
                { ...props }
            />
        );

        wrapper.instance().findType('hsp', 'ijkl9101');

        expect(wrapper.state('hspsToDeselect')).toEqual([ 'ijkl9101' ]);
    });

    it.skip('should remove system from systemsToDeselect', async () => {
        const wrapper = shallow(
            <SelectedBasket
                { ...props }
            />
        );

        wrapper.setState({ systemsToDeselect: [ 'efgh5678' ]});

        await wrapper.instance().findType('system', 'efgh5678');

        expect(wrapper.state('systemsToDeselect')).toEqual([]);
    });

    it('should build array without id', () => {
        const wrapper = shallow(
            <SelectedBasket
                { ...props }
            />
        );

        expect(wrapper.instance().removeId('efgh5678', [ 'efgh5678' ])).toEqual([]);
    });

    it('should build array without id but leave selected', () => {
        const wrapper = shallow(
            <SelectedBasket
                { ...props }
            />
        );

        expect(wrapper.instance().removeId('efgh5678', [ 'abcd1234', 'efgh5678' ])).toEqual([ 'abcd1234' ]);
    });
});
/*eslint-enable camelcase*/
