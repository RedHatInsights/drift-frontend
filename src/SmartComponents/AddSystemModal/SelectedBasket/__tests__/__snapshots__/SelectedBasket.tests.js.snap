// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SelectedBasket should not be visible with selected count 1`] = `
<Fragment>
  <span>
    <Popover
      bodyContent={
        <div
          style={
            Object {
              "maxHeight": "500px",
              "overflowY": "auto",
            }
          }
        >
          <SelectedTable
            findType={[Function]}
            selectedBaselineContent={
              Array [
                Object {
                  "icon": <BlueprintIcon
                    color="currentColor"
                    noVerticalAlign={false}
                    size="sm"
                  />,
                  "id": "abcd1234",
                  "name": "baseline",
                },
              ]
            }
            selectedHSPContent={
              Array [
                Object {
                  "captured_date": "2021-03-03T06:40:32+00:00",
                  "icon": <ClockIcon
                    color="currentColor"
                    noVerticalAlign={false}
                    size="sm"
                  />,
                  "id": "ijkl9101",
                  "system_id": "efgh5678",
                },
              ]
            }
            selectedSystemContent={
              Array [
                Object {
                  "icon": <ServerIcon
                    color="currentColor"
                    noVerticalAlign={false}
                    size="sm"
                  />,
                  "id": "efgh5678",
                  "name": "system",
                },
              ]
            }
          />
        </div>
      }
      footerContent={
        <Button
          onClick={[Function]}
          ouiaId="confirm-selected-basket-button"
          variant="primary"
        >
          Apply changes
        </Button>
      }
      headerContent={
        <div>
          Selected items (
          3
          )
        </div>
      }
      id="selected-basket"
      position="bottom"
      shouldClose={[Function]}
      style={
        Object {
          "minWidth": "500px",
        }
      }
    >
      <a
        onClick={[Function]}
      >
        Selected (
        3
        )
      </a>
    </Popover>
  </span>
</Fragment>
`;

exports[`SelectedBasket should render empty 1`] = `
<Fragment>
  <span>
    <Popover
      bodyContent={
        <div
          style={
            Object {
              "maxHeight": "500px",
              "overflowY": "auto",
            }
          }
        >
          <EmptyStateDisplay
            text={
              Array [
                "Select systems and baselines to compare.",
              ]
            }
            title="Nothing selected"
          />
        </div>
      }
      footerContent={
        <Button
          onClick={[Function]}
          ouiaId="confirm-selected-basket-button"
          variant="primary"
        >
          Apply changes
        </Button>
      }
      headerContent={
        <div>
          Selected items (
          0
          )
        </div>
      }
      id="selected-basket"
      position="bottom"
      shouldClose={[Function]}
      style={
        Object {
          "minWidth": "500px",
        }
      }
    >
      <a
        onClick={[Function]}
      >
        Selected (
        0
        )
      </a>
    </Popover>
  </span>
</Fragment>
`;

exports[`SelectedBasket should toggle to visible with system, baseline and hsp 1`] = `
<Fragment>
  <span>
    <Popover
      bodyContent={
        <div
          style={
            Object {
              "maxHeight": "500px",
              "overflowY": "auto",
            }
          }
        >
          <SelectedTable
            findType={[Function]}
            selectedBaselineContent={
              Array [
                Object {
                  "icon": <BlueprintIcon
                    color="currentColor"
                    noVerticalAlign={false}
                    size="sm"
                  />,
                  "id": "abcd1234",
                  "name": "baseline",
                },
              ]
            }
            selectedHSPContent={
              Array [
                Object {
                  "captured_date": "2021-03-03T06:40:32+00:00",
                  "icon": <ClockIcon
                    color="currentColor"
                    noVerticalAlign={false}
                    size="sm"
                  />,
                  "id": "ijkl9101",
                  "system_id": "efgh5678",
                },
              ]
            }
            selectedSystemContent={
              Array [
                Object {
                  "icon": <ServerIcon
                    color="currentColor"
                    noVerticalAlign={false}
                    size="sm"
                  />,
                  "id": "efgh5678",
                  "name": "system",
                },
              ]
            }
          />
        </div>
      }
      footerContent={
        <Button
          onClick={[Function]}
          ouiaId="confirm-selected-basket-button"
          variant="primary"
        >
          Apply changes
        </Button>
      }
      headerContent={
        <div>
          Selected items (
          3
          )
        </div>
      }
      id="selected-basket"
      position="bottom"
      shouldClose={[Function]}
      style={
        Object {
          "minWidth": "500px",
        }
      }
    >
      <a
        onClick={[Function]}
      >
        Selected (
        3
        )
      </a>
    </Popover>
  </span>
</Fragment>
`;
